src/storage/src/source/reclock.rs

```diff
@@ -8,413 +8,668 @@
-                    )
-                })
-                .collect()
-        } else {
-            HashMap::new()
-        };
-        let read_cursors = HashMap::new();
+        // Ensure we have enough bindings and re-sync state if we lost a race
+        while let Err(Upper(global_upper)) = self.mint(&batch_upper).await {
+            self.sync(&global_upper).await;
+        }
 
-        let timestamp_bindings_listener = read_handle
-            .listen(as_of)
-            .await
-            .expect("Initial listen at handle.since ts")
-            .into_stream()
-            .boxed();
-
-        Ok(Self {
-            name,
-            read_progress,
-            write_upper,
-            persisted_timestamp_bindings,
-            read_cursors,
-            write_handle,
-            timestamp_bindings_listener,
-            now,
-            update_interval,
-            last_timestamp: Timestamp::minimum(),
-        })
+        ReclockIter {
+            reclock: self,
+            messages: batch.iter_mut(),
+        }
     }
 
-    // TODO: require NowFn to be monotonic so we can simplify this function
-    fn get_time(&mut self) -> Option<Timestamp> {
-        let update_interval: u64 = self
-            .update_interval
-            .clone()
-            .as_millis()
-            .try_into()
-            .expect("interval millis don't fit into u64");
-        let new_ts = (self.now)();
-
-        // Mod and then round up
-        let remainder = new_ts % update_interval;
-        let new_ts_rounded = if remainder == 0 {
-            new_ts
-        } else {
-            new_ts - remainder + update_interval
-        };
+    /// Calling this method promises that all future timestamping requests will be for offsets
+    /// greater or equal to the provided source upper frontier. In exchange the reclock operator
+    /// return the timestamp frontier that the caller can use to downgrade its capability to.
+    pub async fn advance(
+        &mut self,
+        source_upper: &HashMap<PartitionId, MzOffset>,
+    ) -> Antichain<Timestamp> {
+        // First ensure that the bindings cover the provided frontier
+        while let Err(Upper(global_upper)) = self.mint(&source_upper).await {
+            self.sync(&global_upper).await;
+        }
 
-        if new_ts_rounded > self.last_timestamp {
-            self.last_timestamp = new_ts_rounded;
-            Some(new_ts_rounded)
-        } else {
-            None
+        // Then, figure out the highest timestamp the caller we downgrade to, which is the minimum
+        // timestamp of all the bindings involved in the given frontier.
+        let mut handle_since = self.upper.clone();
+        for pid in self.last_source_upper.keys() {
+            let offset = source_upper.get(pid).cloned().unwrap_or_default();
+            if let Some((ts, _)) = self
+                .partition_uppers(pid)
+                .find(|(_, upper)| offset < *upper)
```

Comment:
Should this be `offset <= *upper`? I think we'd want to constrain the returned antichain if we're going to allow for more messages at the offset given by `source_upper[pid]` as they'd be reclocked at `ts`.

I don't _think_ there's a test that stresses multiple messages for a given offset but if I'm mistaken can you point me there? If they are in fact missing, can you add some?